crates/wasmparser/src/validator/types.rs

```diff
@@ -0,0 +1,1157 @@
+                    None => false,
+                })
+    }
+}
+
+/// Represents a type of an instance.
+#[derive(Clone)]
+pub struct InstanceType {
+    /// The exports of the instance type.
+    pub exports: HashMap<String, ComponentEntityType>,
+}
+
+impl InstanceType {
+    pub(crate) fn is_subtype_of(&self, other: &Self, types: &TypeList) -> bool {
+        // For instance type subtyping, all exports in the other instance type
+        // must be present in this instance type's exports (i.e. it can export
+        // *more* than what this instance type needs).
+        other
+            .exports
+            .iter()
+            .all(|(k, other)| match self.exports.get(k) {
+                Some(ty) => ty.is_subtype_of(other, types),
+                None => false,
+            })
+    }
+}
+
+/// Represents a type of a component function.
+#[derive(Clone)]
+pub struct ComponentFuncType {
+    /// The function parameters.
+    pub params: Box<[(Option<String>, InterfaceTypeRef)]>,
+    /// The function's result type.
+    pub result: InterfaceTypeRef,
+    pub(crate) core_type: FuncType,
+}
+
+impl ComponentFuncType {
+    pub(crate) fn requires_into_option(&self, types: &TypeList) -> bool {
+        self.result.requires_into_option(types)
+            || self
+                .params
+                .iter()
+                .any(|(_, ty)| ty.requires_into_option(types))
+    }
+
+    pub(crate) fn is_subtype_of(&self, other: &Self, types: &TypeList) -> bool {
+        // Subtyping rules:
+        // https://github.com/WebAssembly/component-model/blob/17f94ed1270a98218e0e796ca1dad1feb7e5c507/design/mvp/Subtyping.md
+
+        // Covariant on return type
+        if !self.result.is_subtype_of(&other.result, types) {
+            return false;
+        }
+
+        // All overlapping parameters must have the same name and are contravariant subtypes
+        for ((name, ty), (other_name, other_ty)) in self.params.iter().zip(other.params.iter()) {
+            if name != other_name {
+                return false;
+            }
+
+            if !other_ty.is_subtype_of(ty, types) {
+                return false;
+            }
+        }
+
+        // All remaining parameters in the supertype must be optional
+        // All superfluous parameters in the subtype are ignored
+        other
+            .params
+            .iter()
+            .skip(self.params.len())
+            .all(|(_, ty)| ty.is_optional(types))
+    }
+}
+
+/// Represents a variant case.
+#[derive(Debug, Clone)]
+pub struct VariantCase {
+    /// The variant case type.
+    pub ty: InterfaceTypeRef,
+    /// The value of the variant to default to.
+    pub default_to: Option<String>,
+}
+
+/// Represents an interface type.
+#[derive(Debug, Clone)]
+pub enum InterfaceType {
+    /// The type is a primitive interface type.
+    Primitive(PrimitiveInterfaceType),
+    /// The type is a record.
+    Record(HashMap<String, InterfaceTypeRef>),
```

Comment:
These types need to be changed as they can no longer use unstable associative collections so that tools that generate interface definitions from the validator's type information can use the correct ordering. We still need fast lookup for subtype checking in the validator, though.

I'm working on that now.

Reply:
Not entirely happy with the new (and only) dependency on `indexmap` from `wasmparser` to solve this, but it is a very minimal dependency at least. Resolving this for now and we can revisit whether or not it's a good thing in the PR review.